lib/ulid/encoder.ex

```diff
@@ -0,0 +1,35 @@
+defmodule Ulid.Encoder do
+  @moduledoc """
+  Handles the encoding of raw Ulids.
+  """
+
+  @encoding "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
+
+  @doc """
+  Encodes a Ulid binary to its textual form.
+
+  ## Examples
+
+      iex> ulid = Ulid.generate_binary()
+      <<1, 98, 114, 124, 105, 203, 130, 43, 45, 123, 70, 66, 93, 91, 180, 93>>
+      iex> Ulid.Encoder.encode(ulid)
+      "01C9S7RTEBG8NJTYT689ENQD2X"
+  """
+  @spec encode(Ulid.raw()) :: {:ok, Ulid.t()} | :error
+  def encode(<<_::unsigned-size(128)>> = bytes) do
+    encode_bytes(<<0::2, bytes::binary>>, <<>>)
+  catch
+    _ -> :error
+  else
+    encoded -> {:ok, encoded}
+  end
+
+  ## PRIVATE FUNCTIONS
+
+  defp encode_bytes(<<index::unsigned-size(5), rest::bitstring>>, acc) do
+    <<_::bytes-size(index), char::bytes-size(1), _::binary>> = @encoding
+    encode_bytes(rest, acc <> char)
+  end
+
+  defp encode_bytes(<<>>, acc), do: acc
```

Comment:
I wonder if hardcoding the function clauses like with decode would be more or less performant than cutting up the binary with a pattern match. I'm guessing more, by a hair.
```
  @crockford_chars "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
  for n <- 0..31 do
    defp encode_bytes(<<unquote(n)::unsigned-size(5), rest::bitstring>>, acc) do
      encode_bytes(rest, acc <> unquote(String.at(@crockford_chars, n)))
    end
  end
  defp encode_bytes(<<>>, acc), do: acc
```

Reply:
Personally, I'm more a fan of `@crockford_alphabet` as the name than `@encoding` as a person who doesn't know nothin' bout nuffin can google "crockford alphabet" and immediately find out that this is from Crockford's base 32 encoding.

Reply:
```
Name                  ips        average  deviation         median         99th %
new encode       383.63 K        2.61 μs  ±3062.17%           2 μs           7 μs
orig encode      260.11 K        3.84 μs  ±1642.05%           3 μs           9 μs

Comparison: 
new encode       383.63 K
orig encode      260.11 K - 1.47x slower
```

Reply:
much quicker